ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/components/stack/StackComponentState.kt

```diff
@@ -0,0 +1,121 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.unit.dp
+import androidx.window.core.layout.WindowWidthSizeClass
+import com.revenuecat.purchases.ui.revenuecatui.components.ComponentViewState
+import com.revenuecat.purchases.ui.revenuecatui.components.ScreenCondition
+import com.revenuecat.purchases.ui.revenuecatui.components.buildPresentedPartial
+import com.revenuecat.purchases.ui.revenuecatui.components.ktx.toPaddingValues
+import com.revenuecat.purchases.ui.revenuecatui.components.ktx.toShape
+import com.revenuecat.purchases.ui.revenuecatui.components.style.StackComponentStyle
+import com.revenuecat.purchases.ui.revenuecatui.data.PaywallState
+
+@JvmSynthetic
+@Composable
+internal fun rememberUpdatedStackComponentState(
+    style: StackComponentStyle,
+    paywallState: PaywallState.Loaded.Components,
+    selected: Boolean,
+): StackComponentState =
+    rememberUpdatedStackComponentState(
+        style = style,
+        isEligibleForIntroOffer = paywallState.isEligibleForIntroOffer,
+        selected = selected,
+    )
+
+@JvmSynthetic
+@Composable
+internal fun rememberUpdatedStackComponentState(
+    style: StackComponentStyle,
+    isEligibleForIntroOffer: Boolean = false,
+    selected: Boolean = false,
+): StackComponentState {
+    val windowSize = currentWindowAdaptiveInfo().windowSizeClass.windowWidthSizeClass
+
+    return remember(style) {
+        StackComponentState(
+            initialWindowSize = windowSize,
+            initialIsEligibleForIntroOffer = isEligibleForIntroOffer,
+            initialSelected = selected,
+            style = style,
+        )
+    }.apply {
+        update(
+            windowSize = windowSize,
+            isEligibleForIntroOffer = isEligibleForIntroOffer,
+            selected = selected,
+        )
+    }
+}
+
+@Stable
+internal class StackComponentState(
+    initialWindowSize: WindowWidthSizeClass,
+    initialIsEligibleForIntroOffer: Boolean,
+    initialSelected: Boolean,
+    private val style: StackComponentStyle,
+) {
+    private var windowSize by mutableStateOf(initialWindowSize)
+    private var isEligibleForIntroOffer by mutableStateOf(initialIsEligibleForIntroOffer)
+    private var selected by mutableStateOf(initialSelected)
+    private val presentedPartial by derivedStateOf {
+        val windowCondition = ScreenCondition.from(windowSize)
+        val componentState = if (selected) ComponentViewState.SELECTED else ComponentViewState.DEFAULT
+
+        style.overrides?.buildPresentedPartial(windowCondition, isEligibleForIntroOffer, componentState)
+    }
+
+    @get:JvmSynthetic
+    val visible by derivedStateOf { presentedPartial?.partial?.visible ?: true }
```

Comment:
Should this default to the style visibility? I don't know if this is something controllable from the dashboard but seems like a better default IMO?
```suggestion
    val visible by derivedStateOf { presentedPartial?.partial?.visible ?: style.visible }
```

Reply:
Oh good that you mention it! This made me realize that `*ComponentStyle`s don't need a `visible` property anymore. I removed them from `StackComponentStyle` and `TextComponentStyle` here: https://github.com/RevenueCat/purchases-android/pull/1993/commits/bf27e456c3073c76a8cc18b1f9632b18588f4e19. Will need to do the same once `ImageComponentStyle` gets an accompanying `ImageComponentState`. 

The reason for this is that a component is always visible, except when a `PresentedPartial` override tells it it's not. This is to support scenarios where a developer wants to hide a component for a certain window size, for instance. That's why only the `PresentedPartial` overrides have a `visible` property, and when that's not set the component is visible.

Reply:
👍 makes sense! Yeah I didn't think it made much sense to control the visibility from the dashboard like that. This makes more sense!